versioned_hdf5/subchunk_map.py

```diff
@@ -36,280 +52,405 @@
-            chunk_stop = min((chunk_idx + 1) * n, d)
-            chunk_subindexes_for_axis.append(
-                (
-                    Slice(chunk_start, chunk_stop, 1),
-                    (),
-                    i_raw - chunk_start,
-                )
-            )
+@cython.cfunc
+def _index_to_mapper(idx, dset_size: hsize_t, chunk_size: hsize_t) -> IndexChunkMapper:
+    """Convert a one-dimensional index, preprocessed by ndindex, to a mapper"""
+    if isinstance(idx, int):
+        return IntegerMapper(idx, dset_size, chunk_size)
+
+    if isinstance(idx, np.ndarray):
+        if idx.ndim != 1:
+            raise NotImplementedError("array index must be 1-dimensional")
+        idx = _maybe_array_idx_to_slice(idx)
+        if isinstance(idx, np.ndarray):
+            return IntegerArrayMapper(idx, dset_size, chunk_size)
+
+    if isinstance(idx, slice):
+        if idx == slice(0, dset_size, 1):
+            return EverythingMapper(dset_size, chunk_size)
         else:
-            raise NotImplementedError(f"index type {type(i)} not supported")
+            return SliceMapper(idx, dset_size, chunk_size)
 
-        chunk_subindexes.append(chunk_subindexes_for_axis)
+    raise NotImplementedError(f"index type {type(idx)} not supported")
 
-    # Handle the remaining suffix axes on which we did not select, we still need to
-    # break them up into chunks.
-    for n, d in zip(suffix_chunk_size, suffix_shape):
-        chunk_slices_gen = (
-            Slice(chunk_idx * n, min((chunk_idx + 1) * n, d), 1)
-            for chunk_idx in range((d + n - 1) // n)
-        )
-        chunk_subindexes.append(
-            [(chunk_slice, chunk_slice.raw, ()) for chunk_slice in chunk_slices_gen]
-        )
+
+@cython.cfunc
+def _maybe_array_idx_to_slice(idx: Any):  # -> NDArray[np_hsize_t] | slice:
+    """Attempt to convert an integer or boolean array index to a slice"""
+    # boolean array indices can be trivially expressed as an integer array
+    if idx.dtype == bool:
+        idx = np.flatnonzero(idx)
+
+    # Don't copy when converting from np.intp to uint64 on 64-bit platforms
+    idx = asarray(idx, np_hsize_t)
+
+    if not idx.flags.writeable:
+        # Cython doesn't support read-only views in pure Python mode,
+        # so we're forced to copy to prevent a crash in the next line.
+        idx = idx.copy()
+
+    idx_v: hsize_t[:] = idx
+    idx_len = len(idx_v)
+    if idx_len == 0:
+        return slice(0, 0, 1)
+
+    start = idx_v[0]
+    if idx_len == 1:
+        return slice(int(start), int(start + 1), 1)
+    if idx_v[1] <= start:  # step <1
+        return idx
+    stop = idx_v[idx_len - 1] + 1
+    step = idx_v[1] - start
+    if idx_len == 2:
+        return slice(int(start), int(stop), int(step))
+
+    if stop2count(start, stop, step) == idx_len:
+        j = start + step + step
+        for i in range(2, idx_len):
+            if idx_v[i] != j:
+                break
+            j += step
+        else:
```

Comment:
functionally equivalent to `if (idx == np.arange(start, stop, step)).all():`, but faster